tests/tests_data/streaming/test_combined.py

```diff
@@ -135,28 +149,683 @@ def __init__(self, start, end):
+        assert torch.equal(next(dataloader_iter), torch.Tensor([1]))
+        assert torch.equal(next(dataloader_iter), torch.Tensor([10]))
+        assert torch.equal(next(dataloader_iter), torch.Tensor([2]))
+        assert torch.equal(next(dataloader_iter), torch.Tensor([3]))
+        assert torch.equal(next(dataloader_iter), torch.Tensor([4]))
+        assert torch.equal(next(dataloader_iter), torch.Tensor([11]))
+        assert torch.equal(next(dataloader_iter), torch.Tensor([5]))
+        assert torch.equal(next(dataloader_iter), torch.Tensor([6]))
+        assert torch.equal(next(dataloader_iter), torch.Tensor([7]))
+        assert torch.equal(next(dataloader_iter), torch.Tensor([12]))
+        assert torch.equal(next(dataloader_iter), torch.Tensor([8]))
+
+    assert dataloader.state_dict() == {
+        "dataset": {
+            "0": {"num_samples_yielded": 9, "num_workers": 1, "batch_size": batch_size},
+            "1": {"num_samples_yielded": 3, "num_workers": 1, "batch_size": batch_size},
+        },
+        "current_epoch": 0,
+        "latest_worker_idx": 0,
+        "num_samples_yielded": {0: [9, 3]},
+    }
+
+
+@pytest.mark.skipif(sys.platform == "win32" or sys.platform == "darwin", reason="too slow in CI")
+def test_combined_dataset_with_dataloader_2_epochs(tmpdir):
+    data_dir_1 = os.path.join(tmpdir, "data_1")
+    data_dir_2 = os.path.join(tmpdir, "data_2")
+    cache_dir_1 = os.path.join(tmpdir, "cache_dir_1")
+    cache_dir_2 = os.path.join(tmpdir, "cache_dir_2")
+
+    os.makedirs(data_dir_1)
+    os.makedirs(data_dir_2)
+    os.makedirs(cache_dir_1)
+    os.makedirs(cache_dir_2)
+
+    cache = Cache(input_dir=str(data_dir_1), chunk_size=2)
+
+    for i in range(10):
+        cache[i] = i
+
+    cache.done()
+    cache.merge()
+
+    cache = Cache(input_dir=str(data_dir_2), chunk_size=2)
+
+    for i in range(10):
+        cache[i] = i + 5
+
+    cache.done()
+    cache.merge()
+
+    dataset1 = StreamingDataset(input_dir=Dir(cache_dir_1, data_dir_1), shuffle=True)
+    dataset2 = StreamingDataset(input_dir=Dir(cache_dir_2, data_dir_2), shuffle=True)
+    dataset = CombinedStreamingDataset(datasets=[dataset1, dataset2], weights=[0.5, 0.5], seed=12345)
+    dataloader = StreamingDataLoader(dataset, num_workers=3, batch_size=2)
+
+    assert dataset1.current_epoch == 1
+    assert dataset2.current_epoch == 1
+
+    batches_1 = []
+    states_1 = []
+    for batch in dataloader:
+        batches_1.append(batch)
+        states_1.append(dataloader.state_dict())
+
+    assert dataset1.current_epoch == 1
+    assert dataset2.current_epoch == 1
+
+    batches_2 = []
+    states_2 = []
+    for batch in dataloader:
+        batches_2.append(batch)
+        states_2.append(dataloader.state_dict())
+    assert dataset1.current_epoch == 2
+    assert dataset2.current_epoch == 2
+
+    assert sum(torch.equal(b1, b2) for b1, b2 in zip(batches_1, batches_2)) != len(batches_1)
+
+    assert states_1 == [
```

Comment:
Can we not assert the state contents programmatically? Listing them plain feels like it is working against the human. I would write single asserts under a for loop.

Reply:
I prefer to keep it this way. It is faster to update if needed but more importantly. it helped me assert all the values were correct.